python/ray/train/lightning/lightning_trainer.py

```diff
@@ -0,0 +1,374 @@
+        self._ddp_strategy_config = {}
+        self._model_checkpoint_config = {}
+        self._ray_dataset_iter_config = {}
+
+    def set_module_class(
+        self, module_class: Type[pl.LightningModule]
+    ) -> "LightningConfigBuilder":
+        """Set up the Pytorch Lightning module class.
+
+        Args:
+            module_class: A subclass of ``pytorch_lightning.LightningModule``
+            that defines your model and training logic. Note that this is a
+            class definition instead of a class instance.
+        """
+        if not isclass(module_class):
+            raise ValueError("'module_class' must be a class, not a class instance.")
+        if not issubclass(module_class, pl.LightningModule):
+            raise ValueError(
+                "'module_class' must be a subclass of 'pl.LightningModule'!"
+            )
+        self._module_class = module_class
+        return self
+
+    def set_module_init_config(self, **kwargs) -> "LightningConfigBuilder":
+        """Set up the initialization argument list of your lightning module."""
+        self._module_init_config.update(**kwargs)
+        return self
+
+    def set_trainer_init_config(self, **kwargs) -> "LightningConfigBuilder":
+        """Set up the configurations of ``pytorch_lightning.Trainer``.
+
+        For valid arguments to pass, please refer to:
+        https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#init.
+        """
+        self._trainer_init_config.update(**kwargs)
+        return self
+
+    def set_trainer_fit_params(self, **kwargs) -> "LightningConfigBuilder":
+        """Setup the parameter lists for ``pytorch_lightning.Trainer.fit()``.
+
+        For valid arguments to pass, please refer to:
+        https://pytorch-lightning.readthedocs.io/en/stable/common/trainer.html#fit.
+        """
+        self._trainer_fit_params.update(**kwargs)
+        return self
+
+    def set_ddp_strategy_config(self, **kwargs) -> "LightningConfigBuilder":
+        """Set up the configurations of ``pytorch_lightning.Trainer``.
+
+        For valid arguments to pass, please refer to:
+        https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.strategies.DDPStrategy.html
+        """
+        self._ddp_strategy_config.update(**kwargs)
+        return self
+
+    def set_model_checkpoint_config(self, **kwargs) -> "LightningConfigBuilder":
+        """Set up the configurations of ``pytorch_lightning.callbacks.ModelCheckpoint``.
+
+        LightningTrainer creates a `ModelCheckpoint` callback based on this config.
+        The AIR checkpointing and logging methods are triggered in that callback.
+        For valid arguments to pass, please refer to:
+        https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.callbacks.ModelCheckpoint.html
+        """
+        self._model_checkpoint_config.update(**kwargs)
+        return self
+
+    def build(self) -> Dict["str", Any]:
```

Comment:
Do we want to follow RLlib's `AlgorithmConfig`? https://docs.ray.io/en/latest/rllib/package_ref/algorithm.html#defining-algorithms-with-the-algorithmconfig-class

It has this builder pattern, but `AlgorithmConfig.build()` will return an actual rllib algo instance - rather than just the dict of configs. It has a separate `AlgorithmConfig.to_dict` method that's used to pass as a Tune `param_space`.

This is a little tangential to the PR, but I'm thinking we can extend this builder pattern to everything in AIR eventually:

```python
trainer = LightningConfig().training(module_class=MyModule).ddp_strategy(...).run_config().build()
```

Reply:
Yes. I think the chained builder is a pretty nice interface implementation. 

But still, I believe it is better to simply return a dictionary for `LightningConfigBuild.build()` for now. We should avoid making the API of LightningTrainer substantially different from that of other AIR trainers.